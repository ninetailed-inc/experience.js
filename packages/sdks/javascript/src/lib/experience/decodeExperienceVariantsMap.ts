export const decodeExperienceVariantsMap = (
  encodedExperienceVariantsMap: string
): Record<string, number> => {
  return encodedExperienceVariantsMap
    .split(',')
    .map((experienceIdWithVariant) => {
      const [experienceId, _variantIndex] = experienceIdWithVariant.split('=');

      const variantIndex = parseInt(_variantIndex);

      if (!experienceId || !variantIndex) {
        return null;
      }

      return { experienceId, variantIndex };
    })
    .filter((x): x is { experienceId: string; variantIndex: number } => !!x)
    .reduce(
      (acc, curr) => ({ ...acc, [curr.experienceId]: curr.variantIndex }),
      {}
    );
};
