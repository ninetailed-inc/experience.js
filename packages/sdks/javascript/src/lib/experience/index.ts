export * from './types';
export {
  EXPERIENCE_TRAIT_PREFIX,
  selectDistribution,
  isExperienceMatch,
  selectVariant as selectExperienceVariant,
  selectHasVariants as selectHasExperienceVariants,
  selectVariants as selectExperienceVariants,
  selectBaselineWithVariants as selectExperienceBaselineWithVariants,
  selectExperience,
  selectActiveExperiments,
} from '@ninetailed/experience.js-shared';
export { decodeExperienceVariantsMap } from './decodeExperienceVariantsMap';
export { makeExperienceSelectMiddleware } from './makeExperienceSelectMiddleware';
export * from './makeChangesModificationMiddleware';
