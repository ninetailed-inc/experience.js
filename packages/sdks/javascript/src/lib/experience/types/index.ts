export type {
  Reference,
  Baseline,
  VariantRef,
  ExperienceConfiguration,
  BaselineWithVariants,
  Distribution,
  ExperienceType,
} from '@ninetailed/experience.js-shared';
