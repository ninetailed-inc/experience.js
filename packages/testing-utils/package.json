{
  "name": "@ninetailed/testing-utils",
  "version": "7.6.0-beta.1",
  "type": "commonjs",
  "license": "BSL-1.1"
}
