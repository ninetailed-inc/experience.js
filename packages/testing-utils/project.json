{
  "name": "testing-utils",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/testing-utils/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/testing-utils",
        "main": "packages/testing-utils/src/index.ts",
        "tsConfig": "packages/testing-utils/tsconfig.lib.json",
        "assets": ["packages/testing-utils/*.md"]
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["packages/testing-utils/**/*.ts"]
      }
    }
  },
  "tags": []
}
