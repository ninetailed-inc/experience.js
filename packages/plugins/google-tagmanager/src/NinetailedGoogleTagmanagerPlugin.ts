import {
  NinetailedAnalyticsPlugin,
  Template,
  SanitizedElementSeenPayload,
} from '@ninetailed/experience.js-plugin-analytics';
import { TrackComponentProperties } from '@ninetailed/experience.js';
import { template } from '@ninetailed/experience.js-shared';

declare global {
  interface Window {
    dataLayer?: unknown[];
  }
}

type NinetailedGoogleTagmanagerPluginOptions = {
  actionTemplate?: string;
  labelTemplate?: string;

  template?: Template;
};

const TEMPLATE_OPTIONS = {
  interpolate: /{{([\s\S]+?)}}/g,
};

export class NinetailedGoogleTagmanagerPlugin extends NinetailedAnalyticsPlugin {
  public name = 'ninetailed:googleTagmanager';

  constructor(
    private readonly options: NinetailedGoogleTagmanagerPluginOptions = {}
  ) {
    super({
      ...options.template,
      event: 'nt_experience',
      ninetailed_variant: '{{selectedVariantSelector}}',
      ninetailed_experience: '{{experience.id}}',
      ninetailed_experience_name: '{{experience.name}}',
      ninetailed_audience: '{{audience.id}}',
      ninetailed_component: '{{selectedVariant.id}}',
    });
  }

  public initialize = () => {
    if (typeof window !== 'undefined') {
      window.dataLayer = window.dataLayer || [];
    }
  };

  protected async onTrackExperience(
    properties: SanitizedElementSeenPayload,
    hasSeenExperienceEventPayload: Record<string, string>
  ): Promise<void> {
    window.dataLayer?.push(hasSeenExperienceEventPayload);
  }

  protected async onTrackComponent(
    properties: TrackComponentProperties
  ): Promise<void> {
    const { variant, audience, isPersonalized } = properties;

    const action = template(
      this.options.actionTemplate || 'Has Seen Experience',
      { component: variant, audience },
      TEMPLATE_OPTIONS.interpolate
    );
    const label = template(
      this.options.labelTemplate ||
        '{{ baselineOrVariant }}:{{ component.id }}',
      {
        component: variant,
        audience,
        baselineOrVariant: isPersonalized ? 'Variant' : 'Baseline',
      },
      TEMPLATE_OPTIONS.interpolate
    );

    window.dataLayer?.push({
      event: action,
      properties: {
        category: 'Ninetailed',
        label,
        nonInteraction: true,
      },
    });
  }
}
