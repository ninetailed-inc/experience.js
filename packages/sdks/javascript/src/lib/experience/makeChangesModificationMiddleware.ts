import { Change, pipe } from '@ninetailed/experience.js-shared';
import { NinetailedPlugin } from '@ninetailed/experience.js-plugin-analytics';

import { RemoveOnChangeListener } from '../utils/OnChangeEmitter';
import { selectPluginsHavingOnChangeEmitter } from '../plugins/selectPluginsHavingOnChangeEmitter';
import { selectPluginsHavingChangesModificationMiddleware } from '../plugins/selectPluginsHavingChangesModificationMiddleware';
import {
  ChangesModificationMiddleware,
  ChangesModificationMiddlewareArg,
} from '../types/interfaces/HasChangesModificationMiddleware';

/**
 * Arguments for creating a changes modification middleware
 */
type CreateChangesModificationMiddlewareArg = {
  plugins: NinetailedPlugin[];
  changes: Change[];
};

/**
 * Arguments for making a changes modification middleware with change notification
 */
type MakeChangesModificationMiddlewareArg =
  CreateChangesModificationMiddlewareArg & {
    onChange: (middleware: ChangesModificationMiddleware) => void;
  };

/**
 * Result of creating a changes modification middleware
 */
interface ChangesModificationMiddlewareResult {
  addListeners: () => void;
  removeListeners: () => void;
  middleware: ChangesModificationMiddleware;
}

/**
 * Creates a pass-through middleware that doesn't modify changes
 */
const createPassThroughMiddleware = (): ChangesModificationMiddleware => {
  return ({ changes }: ChangesModificationMiddlewareArg) => {
    return { changes };
  };
};

/**
 * Creates a middleware function by composing middleware from multiple plugins
 */
function createChangesModificationMiddleware({
  plugins,
  changes,
}: CreateChangesModificationMiddlewareArg): ChangesModificationMiddleware {
  const pluginsWithMiddleware =
    selectPluginsHavingChangesModificationMiddleware(plugins);

  const middlewareFunctions: ChangesModificationMiddleware[] = [];

  for (const plugin of pluginsWithMiddleware) {
    const middleware = plugin.getChangesModificationMiddleware({
      changes,
    });

    if (middleware !== undefined) {
      middlewareFunctions.push(middleware);
    }
  }

  // If no middleware functions were found, return a pass-through middleware
  if (middlewareFunctions.length === 0) {
    return createPassThroughMiddleware();
  }

  // Compose middleware functions using pipe
  return pipe(...middlewareFunctions);
}

/**
 * Creates a changes modification middleware system with change notification
 *
 * This follows the same pattern as the experience selection middleware system
 */
export const makeChangesModificationMiddleware = ({
  plugins,
  onChange,
  changes,
}: MakeChangesModificationMiddlewareArg): ChangesModificationMiddlewareResult => {
  let removeChangeListeners: RemoveOnChangeListener[] = [];

  const pluginsHavingChangeEmitters =
    selectPluginsHavingOnChangeEmitter(plugins);

  const middleware = createChangesModificationMiddleware({
    plugins,
    changes,
  });

  const addListeners = () => {
    removeChangeListeners = pluginsHavingChangeEmitters.map((plugin) => {
      const listener = () => {
        // When a plugin changes, recreate the middleware and notify
        const updatedMiddleware = createChangesModificationMiddleware({
          plugins,
          changes,
        });
        onChange(updatedMiddleware);
      };

      return plugin.onChangeEmitter.addListener(listener);
    });
  };

  // WARNING: This specific implementation using forEach is required.
  // DO NOT replace with for...of or other loop constructs as they will break functionality.
  // The exact reason is uncertain but appears related to the transpiler.
  const removeListeners = () => {
    removeChangeListeners.forEach((removeListener) => removeListener());
  };

  return {
    addListeners,
    removeListeners,
    middleware,
  };
};
