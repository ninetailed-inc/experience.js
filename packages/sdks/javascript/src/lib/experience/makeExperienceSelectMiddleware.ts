import {
  ExperienceConfiguration,
  Profile,
  Reference,
  VariantRef,
  pipe,
} from '@ninetailed/experience.js-shared';

import { NinetailedPlugin } from '../types/NinetailedPlugin';
import { RemoveOnChangeListener } from '../utils/OnChangeEmitter';
import { selectPluginsHavingOnChangeEmitter } from '../plugins/selectPluginsHavingOnChangeEmitter';
import { selectPluginsHavingExperienceSelectionMiddleware } from '../plugins/selectPluginsHavingExperienceSelectionMiddleware';
import { ExperienceSelectionMiddlewareReturnArg } from '../types/interfaces/HasExperienceSelectionMiddleware';

type MakeExperienceSelectMiddlewareArg<Variant extends Reference> = {
  plugins: NinetailedPlugin[];
  experiences: ExperienceConfiguration<Variant>[];
  baseline: Reference;
  profile: Profile | null;
  onChange: () => void;
};

const createPassThroughMiddleware = <Variant extends Reference>() => {
  return ({
    experience,
    variant,
    variantIndex,
  }: ExperienceSelectionMiddlewareReturnArg<Variant>) => {
    return { experience, variant, variantIndex };
  };
};

export const makeExperienceSelectMiddleware = <Variant extends Reference>({
  plugins,
  onChange,
  experiences,
  baseline,
  profile,
}: MakeExperienceSelectMiddlewareArg<Variant>) => {
  let removeChangeListeners: RemoveOnChangeListener[] = [];

  const pluginsHavingChangeEmitters =
    selectPluginsHavingOnChangeEmitter(plugins);

  const prepareMiddleware = () => {
    if (profile === null) {
      return createPassThroughMiddleware<Variant | VariantRef>();
    }

    const pluginsWithMiddleware =
      selectPluginsHavingExperienceSelectionMiddleware<Variant | VariantRef>(
        plugins
      );

    const middlewareFunctions = pluginsWithMiddleware.map((plugin) =>
      plugin.getExperienceSelectionMiddleware({ experiences, baseline })
    );

    return pipe(...middlewareFunctions);
  };

  const addListeners = () => {
    removeChangeListeners = pluginsHavingChangeEmitters.map((plugin) => {
      const listener = () => {
        onChange();
      };

      return plugin.onChangeEmitter.addListener(listener);
    });
  };

  const removeListeners = () => {
    removeChangeListeners.forEach((listener) => listener());
  };

  const middleware = prepareMiddleware();

  return {
    addListeners,
    removeListeners,
    middleware,
  };
};
