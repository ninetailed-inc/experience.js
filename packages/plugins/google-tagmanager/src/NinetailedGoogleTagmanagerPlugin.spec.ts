import { Ninetailed } from '@ninetailed/experience.js';
import { Template } from '@ninetailed/experience.js-plugin-analytics';
import { fixtures } from '@ninetailed/experience.js-plugin-analytics/test';
import { sleep } from '@ninetailed/testing-utils/sleep';

import { NinetailedGoogleTagmanagerPlugin } from './NinetailedGoogleTagmanagerPlugin';

const setup = (template: Template = {}) => {
  const googleTagmanagerPlugin = new NinetailedGoogleTagmanagerPlugin({
    template,
  });
  const ninetailed = new Ninetailed(
    { clientId: 'test' },
    { plugins: [googleTagmanagerPlugin] }
  );

  return { ninetailed, googleTagmanagerPlugin };
};

describe('NinetailedGoogleTagmanagerPlugin', () => {
  let ninetailed: Ninetailed;

  beforeEach(() => {
    window.dataLayer = [];
    ({ ninetailed } = setup());
  });

  it('should be defined', () => {
    expect(NinetailedGoogleTagmanagerPlugin).toBeDefined();
  });

  it('should be instantiable', () => {
    const plugin = new NinetailedGoogleTagmanagerPlugin();

    expect(plugin).toBeInstanceOf(NinetailedGoogleTagmanagerPlugin);
  });

  it('should add the has_seen_element events to the window.dataLayer', async () => {
    await ninetailed.trackComponentView(
      fixtures.TRACK_COMPONENT_VIEW_PROPERTIES
    );

    await sleep(5);
    expect(window.dataLayer).toHaveLength(1);
    expect(window.dataLayer && window.dataLayer[0]).toMatchSnapshot();
  });

  it('should add the has_seen_component events to the window.dataLayer', async () => {
    await ninetailed.trackHasSeenComponent(fixtures.TRACK_COMPONENT_PROPERTIES);

    await sleep(5);
    expect(window.dataLayer).toHaveLength(1);
    expect(window.dataLayer && window.dataLayer[0]).toEqual({
      event: 'Has Seen Experience',
      properties: {
        category: 'Ninetailed',
        label: 'Variant:test',
        nonInteraction: true,
      },
    });
  });

  describe('template options', () => {
    it('should not override predefined keys', async () => {
      ({ ninetailed } = setup({ ninetailed_experience: 'myKey' }));

      await ninetailed.trackComponentView(
        fixtures.TRACK_COMPONENT_VIEW_PROPERTIES
      );

      await sleep(5);
      expect(window.dataLayer).toHaveLength(1);
      expect(
        window.dataLayer && (window.dataLayer[0] as any).ninetailed_experience
      ).toEqual(fixtures.TRACK_COMPONENT_VIEW_PROPERTIES.experience.id);
    });
  });
});
