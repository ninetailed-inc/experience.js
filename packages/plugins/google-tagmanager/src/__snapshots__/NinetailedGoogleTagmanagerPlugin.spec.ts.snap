// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NinetailedGoogleTagmanagerPlugin should add the has_seen_element events to the window.dataLayer 1`] = `
Object {
  "event": "nt_experience",
  "ninetailed_audience": "test",
  "ninetailed_component": "test",
  "ninetailed_experience": "test",
  "ninetailed_experience_name": "test",
  "ninetailed_variant": "variant 1",
}
`;
